src/anyio/_backends/_asyncio.py

```diff
@@ -1724,62 +1729,87 @@ def _raise_async_exceptions(self) -> None:
+            tuple[Coroutine[Any, Any, T_Retval], Future[T_Retval]]
+        ],
+    ) -> None:
+        with receive_stream:
+            async for coro, future in receive_stream:
+                try:
+                    retval = await coro
+                except BaseException as exc:
+                    if not future.cancelled():
+                        future.set_exception(exc)
+                else:
+                    if not future.cancelled():
+                        future.set_result(retval)
+
+    async def _call_in_runner_task(
+        self, func: Callable[..., Awaitable[T_Retval]], *args: object, **kwargs: object
+    ) -> T_Retval:
+        if not self._runner_task:
+            self._send_stream, receive_stream = create_memory_object_stream(1)
+            self._runner_task = self._loop.create_task(
+                self._run_tests_and_fixtures(receive_stream)
+            )
+
+        coro = func(*args, **kwargs)
+        future: asyncio.Future[T_Retval] = self._loop.create_future()
+        self._send_stream.send_nowait((coro, future))
+        return await future
+
     def close(self) -> None:
         try:
             self._cancel_all_tasks()
             self._loop.run_until_complete(self._loop.shutdown_asyncgens())
         finally:
             asyncio.set_event_loop(None)
             self._loop.close()
+            self._runner_task = None
 
     def run_asyncgen_fixture(
         self,
         fixture_func: Callable[..., AsyncGenerator[T_Retval, Any]],
         kwargs: dict[str, Any],
     ) -> Iterable[T_Retval]:
-        async def fixture_runner() -> None:
-            agen = fixture_func(**kwargs)
-            try:
-                retval = await agen.asend(None)
-                self._raise_async_exceptions()
-            except BaseException as exc:
-                f.set_exception(exc)
-                return
-            else:
-                f.set_result(retval)
-
-            await event.wait()
-            try:
-                await agen.asend(None)
-            except StopAsyncIteration:
-                pass
-            else:
-                await agen.aclose()
-                raise RuntimeError("Async generator fixture did not stop")
-
-        f = self._loop.create_future()
-        event = asyncio.Event()
-        fixture_task = self._loop.create_task(fixture_runner())
-        self._loop.run_until_complete(f)
-        yield f.result()
-        event.set()
-        self._loop.run_until_complete(fixture_task)
+        asyncgen = fixture_func(**kwargs)
+        fixturevalue: T_Retval = self._loop.run_until_complete(
+            self._call_in_runner_task(asyncgen.asend, None)
+        )
         self._raise_async_exceptions()
 
+        yield fixturevalue
+
+        try:
+            self._loop.run_until_complete(
+                self._call_in_runner_task(asyncgen.asend, None)
+            )
+        except StopAsyncIteration:
+            self._raise_async_exceptions()
```

Comment:
Why is this reraised now? It wasn't previously and is not in the trio implementation.

Reply:
Trio does not have callbacks like asyncio does. Here we raise any errors from those callbacks that occurred during the run. I'm unsure if we should do this in the setup phase too.

Reply:
Now that I look at the previous code, we did that in the setup phase but not teardown. I think it might make sense to do it in both.

Reply:
Ah I see it's raising exceptions that occurred in teardown callbacks not reraising the captured iteration error. Might be nice to include a brief comment noting that.

Raising exceptions during setup and teardown makes sense right? Is there a reason you'd ever want them to pass silently? I wish it was easier to see the expected behavior in these tests https://github.com/agronholm/anyio/blob/trampoline-test/tests/test_pytest_plugin.py#L155-L159

Reply:
I think a clarification is necessary. What's being raised here are errors caught by the asyncio exception handler, for callbacks scheduled via `loop.call_soon()`, `loop.call_later()` and `loop.call_at()`, not pytest's setup/teardown callbacks.

And as for reraising these in setup/teardown, I forgot that I'm already raising them in both the setup/teardown phases as we should be doing.

Reply:
Great thanks for taking the time to explain! I should have just read the implementation of `_raise_async_exceptions` in the first place but this was helpful.